### ISSUE TITLE
Develop a lightweight client for Rucio

### ISSUE THREAD

**Issue by jamesp-epcc** [rucio/rucio#4097]
It would be desirable to have a lightweight version of the Rucio client with fewer dependencies to make installation by end users easier. Mario suggests the following things need to happen for this:

1. Get rid of libraries used for convenience and instead use Python standard libraries (e.g. network interaction, printing out, progress bar, tabulate, dogpile, etc.).
2. Remove server side mapping of library to be used for access to each RSE, instead make this a client side option so client can use (e.g.) HTTP or Xrootd instead of GFAL.
3. Client should dynamically check its environment to see what libraries are available for accessing the storage.

Questions to be resolved:

1. **Should the lightweight client be a replacement for the standard Rucio client, or an additional standalone alternative client?** If we can do this without too much disruption and still support all of the required features, it could be a modification to the standard client. This would avoid having to maintain two separate client code bases and keep them in sync.
2. **Which features does the lightweight client need to support?** Upload, download and list replicas are definitely required. Upload and download are the difficult ones since they involve direct interaction with the storage; other features are mostly simple REST calls to the server so it might be feasible to support them all. On the other hand, if the lightweight client is standalone, fewer features would be easier to keep in sync with the main client.
3. **Should this be Python 3 only, or also support Python 2.7?** Dropping Python 2 would allow us to rely on the richer features available in the Python 3 standard library (e.g. urllib to replace requests), simplifying the code, but would limit the environments in which the client could run, which may be unacceptable since the main purpose of this change is to allow the client to easily be installed anywhere.

As a first step I will study the relevant code in detail and come up with a plan (or maybe multiple options for further discussion).

**Comment by mlassnig** [comment]
Hi James, thanks. I think the ideal way would be to modify the client directly, and phase out usage of existing dependencies over time, because there's no reason to keep a "heavy" client in any case. This would simplify everyone's life. For Py2/3 I have a slight tendency to drop support for Py2 in the client, but we should maybe see how much we can already drop and still stick to Py2.

**Comment by jamesp-epcc** [comment]
I have looked at the client code to see where all the current dependencies are used. I thought I would write up my ideas on how to deal with each one. Please let me know if you have other suggestions.

- `setuptools` - only used in `setup.py`. I guess we need to keep this, but if we make the client Py3 only we could probably get rid of the upper version limit.
- `argcomplete` - can't remove this without losing tab completion of RSEs, but could make it optional.
- `requests` - used extensively. One of the harder ones to deal with. Could replace with Py3 urllib functionality.
- `urllib3` - only used to disable warnings and force a specific TLS version, could probably do those things some other way.
- `dogpile.cache` - used quite a lot. Could replace with a very simple caching system within Rucio, implementing just the essential features.
- `boto` - only used in the s3boto protocol. Could either get rid of this legacy protocol altogether and just use HTTP with signed URLs, or make the library optional.
- `tabulate` - replace with custom code.
- `progressbar` - replace with custom code.
- `bz2file` - only used in `test_dumper.py`. Could make this optional, only needed for people who actually want to run the tests.
- `python_magic` - doesn't appear to be used in client.
- `futures` - doesn't appear to be used in client.
- `six` - used for simple string type checking, conversion and version checking. Replace with custom code.
- `boto3` - not used in client.
- `pysftp` - used in SFTP protocol. Make this optional and autodetect it.
- `jsonschema` - only used for `validate_schema` function, which is never called from client. It is used in the tests however. Could make optional.

In summary, I think it would be manageable to remove the majority of these and make several others optional.

**Comment by bziemons** [comment]
`six` is used for Python 2/3 compatibility code, so that could be removed when moving to py3-only.

Just to give my 2 cents: I strongly oppose replacing well functioning and available python packages with custom code. `requests` for example has very few dependencies itself and builds upon Python's standard `urllib`. Replacing it and packages alike will require significantly more development effort *also* in the long run. At the same time I strongly encourage removing unused dependencies and make dependencies optional where feasible.

**Comment by mlassnig** [comment]
`requests` is a bit problematic, because especially on RHEL machines this is an OS-level dependency and the clients would be installed OS-wide. We are having quite a lot of problems with these sites not being able to upgrade the clients because changing the `requests` version would break OS functionality. Maybe one solution for this would be to bundle `requests` directly with the clients instead of depending on it (as a last measure)?

**Comment by bziemons** [comment]
@mlassnig is this really an issue? Aren't we just fine with that bigger range of versions of `requests` or do they change their API all the time?

in other words: I think this is more an issue in how we define dependencies. As in `==` instead of a range for the versions that are supported by our target OSes.

**Comment by bari12** [comment]
@jamesp-epcc it would be good to have somewhat of a motivational summary which dependencies (from the POV of DUNE) are the problematic ones. For the point (1) of your issue description. My suspicion is that while it is very nice to get rid of dependencies like `progressbar` or `python_magic` that these ones are not problematic anyway.

I would guess that this probably boils down to `requests`, but I agree with @bziemons duplicating (and supporting!) the functionality of one of the most used packages in python should be very well motivated. We should investigate if the underlying issues cannot be solved otherwise.

**Comment by jamesp-epcc** [comment]
(Apologies for the lateness of this response, I had to take time off unexpectedly due to family circumstances, so I am a bit behind with everything).

We discussed this at a DUNE meeting yesterday. The general feeling from people who have actually used the client in a DUNE environment was that the number of dependencies was more of a problem than any specific individual dependency. Ideally they would like a Rucio client with no external dependencies, however they recognise that this may be an unrealistic goal. They think they can persuade the maintainers of the DUNE Python environment to include `requests`, since it is very useful, but probably not all of the lesser-used libraries.

They mentioned having had problems with GFAL, although GFAL is not actually listed as a hard dependency for the client. I think the problem was that their system GFAL wouldn't work with the Python environment so they had to build it from source instead, which was quite a lot of work. It would probably improve matters to have the client check what libraries are available for RSE access and use whatever it finds, as we discussed previously.

It sounds as if everyone is in agreement that it would be good to at least remove the unnecessary dependencies and make some others optional, so I can start with that.

**Comment by gabrielefronze** [comment]
Hi @jamesp-epcc , all,
we would be quite interested on this project.
Having such light weight client copes really well with the fuse module, which right now works well only on CentOS due to GFAL non being distributed e.g. on macos.
I kind of agree with the others about removing packages and replacing them with custom code, but I’d gladly help on getting rid of GFAL for the reasons posted above.
Count me in!

**Comment by paulmillar** [comment]
@jamesp-epcc This sounds like a good exercise and I agree with @mlassnig that trying to update the standard client (as much as possible) would be a good way to go.

I don't have much to contribute, but to as a comment about your second point (`Remove server side mapping...`).

I wasn't sure if the idea here is simply dropping a dependency on gfal, or whether you wanted to by-pass (or avoid altogether) asking Rucio with which protocol data should be transferred.

gfal is meant as a harmonising layer that makes it easier to write clients that interact with storage by hiding the details on how to transfer data.  AFAIK, there's no need for the rucio client to use it -- however, dropping gfal dependency would move the complexity (currently hidden by gfal) into the rucio client.  Doing so makes the rucio client easier to deploy, but also harder to maintain.  I'm not saying "keep gfal", but just to make sure the consequences are clear.

In terms of choosing which protocol to use for transfers, that's probably a good idea, but users may want to use a "reasonable default" supplied by the Rucio server.

**Comment by mlassnig** [comment]
Hi Paul,

the objective is not to remove gfal, but to make rucio-client smarter at runtime where it can detect which tools are available (e.g., if xrd-utils or gfal-utils or curl etc are installed), and then make a good decision which of the tools to use.
